core/modules/loader/DoList.h

```diff
@@ -0,0 +1,195 @@
+    using Clock = std::chrono::system_clock;
+
+    TimeOut(std::chrono::milliseconds timeOut) : _timeOut(timeOut) {}
+
+    bool due() { return due(Clock::now()); }
+    bool due(TimePoint now) {
+        auto triggerDiff = std::chrono::duration_cast<std::chrono::milliseconds>(now - _lastTrigger);
+        return (triggerDiff > _timeOut);
+    }
+
+    void triggered() { return triggered(Clock::now()); }
+    void triggered(TimePoint now) {
+        _lastTrigger = now;
+    }
+
+    std::chrono::milliseconds timeLeft(TimePoint now) {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(now - _lastTrigger);
+    }
+
+    void setTimeOut(std::chrono::milliseconds timeOut) { _timeOut = timeOut; }
+    std::chrono::milliseconds getTimeOut() const { return _timeOut; }
+private:
+    // How much time since lastTrigger needs to pass before triggering.
+    std::chrono::milliseconds _timeOut; // default, 15 minutes
+    TimePoint _lastTrigger{std::chrono::seconds(0)};
+};
+
+
+/// Children of this class must be created with shared pointers.
+class DoListItem : public std::enable_shared_from_this<DoListItem> {
+public:
+    using Ptr = std::shared_ptr<DoListItem>;
+
+    DoListItem() = default;
+
+    DoListItem(DoListItem const&) = delete;
+    DoListItem& operator=(DoListItem const&) = delete;
+
+    virtual ~DoListItem() = default;
+
+    util::CommandTracked::Ptr runIfNeeded(TimeOut::TimePoint now) {
+        std::lock_guard<std::mutex> lock(_mtx);
+        if (_command == nullptr) {
+            if (_isOneShotDone()) { return nullptr; }
+            if ((_needInfo || _timeOut.due(now)) && _timeRequest.due(now)) {
+                _timeRequest.triggered();
+                    _command = createCommand();
+                return _command;
+            }
+        } else if (_command->isFinished()) {
+            _command.reset(); // Allow the command to be sent again later.
+        }
+        return nullptr;
+    }
+
+    bool isAlreadyOnList() { return _addedToList; }
+
+    /// Returns original value of _addedToList.
+    bool setAddedToList(bool value) {
+        return _addedToList.exchange(value);
+    }
+
+    bool removeFromList() {
+        std::lock_guard<std::mutex> lock(_mtx);
+        return (_isOneShotDone() || _remove);
+    }
+
+    /// The info has been updated, so no need to ask for it for a while.
+    void infoReceived() {
+        std::lock_guard<std::mutex> lock(_mtx);
+        _needInfo = false;
+        _timeOut.triggered();
+    }
+
+    void setNeedInfo() {
+        std::lock_guard<std::mutex> lock(_mtx);
+        _needInfo = true;
+    }
+
+    DoListItem::Ptr getDoListItemPtr() {
+        return shared_from_this();
+    }
+
+    void setTimeOut(std::chrono::milliseconds timeOut) { _timeOut.setTimeOut(timeOut); }
+
+    virtual util::CommandTracked::Ptr createCommand()=0;
+
+protected:
+    std::atomic<bool>    _addedToList{false}; ///< True when added to a DoList
```

Comment:
I think protected members should NOT begin with `_`. Here is the link to the relevant section of the Style guide:
https://developer.lsst.io/cpp/style.html#the-parts-of-a-class-must-be-sorted-public-protected-and-private

And this is for `private` members:
https://developer.lsst.io/cpp/style.html#private-class-variables-and-methods-must-have-underscore-prefix